core/src/ephemeral_server/mod.rs

```diff
@@ -0,0 +1,461 @@
+    let temp_dest = Path::new(&temp_dest_str);
+    // This match only gets Ok if the file was downloaded and extracted to the
+    // temporary path
+    match new_executable_file(temp_dest) {
+        Err(err) if err.kind() == std::io::ErrorKind::AlreadyExists => {
+            // Since it already exists, we'll try once a second for 20 seconds
+            // to wait for it to be done, then recurse
+            for _ in 0..20 {
+                sleep(Duration::from_secs(1)).await;
+                if !temp_dest.exists() {
+                    return Ok(false);
+                }
+            }
+            Err(anyhow!(
+                "Temp download file at {} not complete after 20 seconds",
+                temp_dest.display()
+            ))
+        }
+        Err(err) => Err(err.into()),
+        // If the dest was added since, just remove temp file
+        Ok(_) if dest.exists() => {
+            std::fs::remove_file(temp_dest)?;
+            return Ok(true);
+        }
+        // Download and extract the binary
+        Ok(mut temp_file) => {
+            info!("Downloading {} to {}", uri, dest.display());
+            download_and_extract(client, uri, file_to_extract, &mut temp_file)
+                .await
+                .map_err(|err| {
+                    // Failed to download, just remove file
+                    if let Err(err) = std::fs::remove_file(temp_dest) {
+                        warn!(
+                            "Failed removing temp file at {}: {:?}",
+                            temp_dest.display(),
+                            err
+                        );
+                    }
+                    err
+                })
+        }
+    }?;
+    // Now that file should be dropped, we can rename
+    std::fs::rename(temp_dest, dest)?;
+    Ok(true)
+}
+
+#[cfg(target_family = "unix")]
+fn new_executable_file(file: &Path) -> std::io::Result<File> {
+    OpenOptions::new()
+        .create_new(true)
+        .write(true)
+        .mode(0o755)
+        .open(file)
+}
+
+#[cfg(not(target_family = "unix"))]
+fn new_executable_file(file: &Path) -> std::io::Result<File> {
+    OpenOptions::new().create_new(true).write(true).open(file)
+}
+
+async fn download_and_extract(
+    client: &reqwest::Client,
+    uri: &str,
+    file_to_extract: &Path,
+    dest: &mut std::fs::File,
+) -> anyhow::Result<()> {
+    // Start download. We are using streaming here to extract the file from the
+    // tarball or zip instead of loading into memory for Cursor/Seek.
+    let resp = client.get(uri).send().await?;
+    // We have to map the error type to an io error
+    let stream = resp
+        .bytes_stream()
+        .map(|item| item.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err)));
+
+    // Since our tar/zip impls use sync IO, we have to create a bridge and run
+    // in a blocking closure. We have to extend the lifetime of some references
+    // we are given for this.
```

Comment:
```suggestion
    // in a blocking closure.
```
Not totally accurate (you don't really extend any lifetimes here, that isn't really possible in safe code - it's just a matter of shifting ownership), but I think fine to leave off this bit of comment anyway.

Reply:
Can leave it off. Yeah, I didn't mean whatever is technically meant by "extending lifetimes", I just meant I had to do things to make these live longer.

Reply:
Done